presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java

```diff
@@ -321,7 +334,102 @@ private static int hashBytes(int initialValue, Slice bytes)
+            Map<String, NullableValue> currentBinding,
+            List<ColumnHandle> columnHandles,
+            int index,
+            List<Pair<String, TypeInfo>> bucketColTypeInfo,
+            int bucketCount,
+            ImmutableSet.Builder<Integer> builder)
+    {
+        if (index == columnHandles.size()) {
+            OptionalInt singleBucket = getHiveBucket(bucketCount, bucketColTypeInfo, currentBinding);
+            if (singleBucket.isPresent()) {
+                builder.add(singleBucket.getAsInt());
+                return true;
+            } else {
+                return false;
+            }
+        }
+        else {
+            ColumnHandle handle = columnHandles.get(index);
+            for (NullableValue val : bindings.get(handle)) {
+                currentBinding.put(((HiveColumnHandle)handle).getName(), val);
+                if (!getHiveBucketsHelper(bindings, currentBinding, columnHandles, index + 1, bucketColTypeInfo, bucketCount, builder)) {
+                    return false;
+                };
+            }
+            return true;
+        }
+    }
+
+    private static OptionalInt getHiveBucket(int bucketCount, List<Pair<String, TypeInfo>> bucketColumnTypeInfos, Map<String, NullableValue> bindings)
+    {
+        if (bindings.isEmpty()) {
+            return OptionalInt.empty();
+        }
+
+        // Check that we have bindings for all bucket columns
+        if (bindings.size() != bucketColumnTypeInfos.size()) {
+            return OptionalInt.empty();
+        }
+
+        ImmutableList.Builder<TypeInfo> typeInfos = ImmutableList.builder();
+        Object[] values = new Object[bucketColumnTypeInfos.size()];
+        for (int i=0; i<bucketColumnTypeInfos.size(); i++) {
+            Pair<String, TypeInfo> item = bucketColumnTypeInfos.get(i);
+            // All bucket columns must have bindings
+            if (!bindings.containsKey(item.getLeft()) || bindings.get(item.getLeft()).isNull()) {
+                return OptionalInt.empty();
+            }
+            typeInfos.add(item.getRight());
+            values[i] =  bindings.get(item.getLeft()).getValue();
+        }
+
+        return OptionalInt.of(getHiveBucket(bucketCount, typeInfos.build(), values));
+    }
+
+    private static Optional<List<Pair<String, TypeInfo>>> getBucketColumnTypeInfos(Table table)
+    {
+        List<String> bucketColumns = table.getStorage().getBucketProperty().get().getBucketedBy();
+        Map<String, HiveType> hiveTypes = new HashMap<>();
+        for (Column column : table.getDataColumns()) {
+            hiveTypes.put(column.getName(), column.getType());
+        }
+
+        // Verify the bucket column types are supported
+        for (String column : bucketColumns) {
+            if (!SUPPORTED_TYPES_FOR_BUCKET_FILTER.contains(hiveTypes.get(column))) {
+                return Optional.empty();
+            }
+        }
+
+        List<Pair<String, TypeInfo>> columnTypeInfoPairs = new ArrayList<>();
```

Comment:
this logic can be simplified using streaming APIs:

```
    private static Optional<List<Pair<String, TypeInfo>>> getBucketColumnTypeInfos(Table table)
    {
        List<String> bucketColumns = table.getStorage().getBucketProperty().get().getBucketedBy();

        // Verify the bucket column types are supported
        Map<String, HiveType> hiveTypes = table.getDataColumns().stream()
                .collect(toImmutableMap(Column::getName, Column::getType));
        for (String column : bucketColumns) {
            if (!SUPPORTED_TYPES_FOR_BUCKET_FILTER.contains(hiveTypes.get(column))) {
                return Optional.empty();
            }
        }

        return Optional.of(bucketColumns.stream()
                .map(column -> new ImmutablePair<>(column, hiveTypes.get(column).getTypeInfo()))
                .collect(toImmutableList()));
    }
```

Reply:
Thanks! Done